src/main.rs

```diff
@@ -1,3 +1,203 @@
+use std::io::{self, prelude::*};
+use std::os::unix::net::UnixStream;
+use std::process::{Command, Stdio};
+use std::{thread, time};
+
+#[repr(u8)]
+enum MsgType {
+    MsgQuit = 1,
+    MsgRunProcess,
+}
+
+#[repr(u8)]
+enum MsgSubType {
+    SubMsgEnd = 0,
+    SubMsgRunProcessBin,
+    SubMsgRunProcessArg,
+    SubMsgRunProcessEnv,
+    SubMsgRunProcessUid,
+    SubMsgRunProcessRfd,
+}
+
+#[repr(u8)]
+#[derive(Debug)]
+enum RespType {
+    RespOk = 0,
+    RespErr,
+    NotifyOutputAvailable,
+    NotifyProcessDied,
+}
+
+impl RespType {
+    fn from_u8(x: u8) -> io::Result<Self> {
+        match x {
+            0 => Ok(RespType::RespOk),
+            1 => Ok(RespType::RespErr),
+            2 => Ok(RespType::NotifyOutputAvailable),
+            3 => Ok(RespType::NotifyProcessDied),
+            _ => Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Invalid response type",
+            )),
+        }
+    }
+}
+
+struct GuestAgent {
+    stream: UnixStream,
+    last_msg_id: u64,
+}
+
+impl GuestAgent {
+    fn connected(path: &str) -> io::Result<GuestAgent> {
```

Comment:
Shouldn't this be named `connect` ?

Reply:
This is intentional (however might not be the best name). `connect` would mean that this is a method used to connect, not a constructor.